SPEC.md

```diff
@@ -77,33 +78,58 @@ Network configuration in JSON format is streamed to the plugin through stdin. Th
 
 ### Result
 
-Success is indicated by a return code of zero and the following JSON printed to stdout in the case of the ADD command. This should be the same output as was returned by the IPAM plugin (see [IP Allocation](#ip-allocation) for details).
+Success is indicated by a return code of zero and the following JSON printed to stdout in the case of the ADD command. The `ip` and `dns` items should be the same output as was returned by the IPAM plugin (see [IP Allocation](#ip-allocation) for details) except that the plugin should fill in the `interface` indexes appropriately, which are missing from IPAM plugin output since IPAM plugins should be unaware of interfaces.
 
 ```
 {
   "cniVersion": "0.1.0",
-  "ip4": {
-    "ip": <ipv4-and-subnet-in-CIDR>,
-    "gateway": <ipv4-of-the-gateway>,  (optional)
-    "routes": <list-of-ipv4-routes>    (optional)
-  },
-  "ip6": {
-    "ip": <ipv6-and-subnet-in-CIDR>,
-    "gateway": <ipv6-of-the-gateway>,  (optional)
-    "routes": <list-of-ipv6-routes>    (optional)
-  },
+  "interfaces": [
```

Comment:
@dcbw While I don't view that all of container networking functionality needs to be in CNI (only parts that are meant to be pluggable), your use cases are certainly valid (in a sense that they need to be implemented somehow).

But the part that I'm missing is how will you set up traffic shaping if the plugin can return whatever it wants in the `interfaces` section. Will you try to introspect the "host" interfaces to see what they are and where they might be connected to?

Specific comments:
> host port<->container port mapping which, if you're using Docker networking you get for free but of course this is CNI so you have to invent it yourself. We've done that in Kubernetes but the iptables rules normally used for this require the host-side bridge interface name. I suppose we could write a CNI port forwarding plugin and chain it but that would still require the host bridge interface name.

rkt implements port mapping via DNAT to container-ip, assuming that it's reachable from the host.

> traffic shaping which typically requires the host bridge interface name and the host-side veth interface name. Could be done as a CNI plugin or utility library too, but still needs chaining.

I know very little about traffic shaping but I would think you'd want to do that to the interface inside the container. For example, I doubt you can set tc on a slave of macvlan from the host side.